python/pyspark/sql/tests/test_dataframe.py

```diff
@@ -63,6 +62,51 @@
 
 
 class DataFrameTestsMixin:
+    def test_getitem_invalid_indices(self):
+        df = self.spark.sql(
+            "SELECT * FROM VALUES "
+            "(1, 1.1, 'a'), "
+            "(2, 2.2, 'b'), "
+            "(4, 4.4, 'c') "
+            "AS TAB(a, b, c)"
+        )
+
+        # accepted type and values
+        for index in [False, True, 0, 1, 2, -1, -2, -3]:
+            df[index]
```

Comment:
This is really a bad API. `df.col` can be ambiguous as people may use the column reference far away from the dataframe, e.g. `df1.join(df2).select...filter...select(df1.col)`. We recommend users use qualified unresolved column instead, like `col("t1.col")`. Now `df[index]` is even worse as it only makes sense to use it intermediately in current df's transformation.

Why do we add such an API? To support order by ordinal, we can just order by integer literals. The SQL parser also parses `ORDER BY 1, 2` as ordering by integer literal 1 and 2, and analyzer will properly resolve it.

cc @HyukjinKwon @zhengruifeng

Reply:
If `df[index]` is already in pyspark for a while, I think it's fine to treat it as a shortcut of `df.i_th_col`. We shouldn't use `GetColumnByOrdinal` in this case, as it was added for Dataset Tuple encoding and it's guaranteed that we want to get the column from the direct child of the current plan node. But here, we can't guarantee this, as people can do `df1.select..filter...groupBy...select(df1[index])`

Reply:
https://github.com/apache/spark/commit/6183b5e2caedd074073d0f6cb6609a634e2f5194

`df[index]` has been supported since spark 2.0.0.

To support `df.groupBy(1, 2, 3)` and `df.orderBy(1, 2, 3)`, right now `GetColumnByOrdinal` is only used in the direct child internally.

> The SQL parser also parses ORDER BY 1, 2 as ordering by integer literal 1 and 2, and analyzer will properly resolve it.

Do you mean use should directly `SortOrder(UnresolvedOrdinal(index))` ?

Reply:
have offline discussion with wenchen, will fix it by switching to `SortOrder(Literal(index))`. Will fix it next week.

Reply:
```
scala> val df = Seq((2, 1), (1, 2)).toDF("a", "b")
val df: org.apache.spark.sql.DataFrame = [a: int, b: int]

scala> df.show()
+---+---+
|  a|  b|
+---+---+
|  2|  1|
|  1|  2|
+---+---+


scala> df.orderBy(lit(1)).show()
+---+---+
|  a|  b|
+---+---+
|  1|  2|
|  2|  1|
+---+---+


scala> df.groupBy(lit(1)).agg(first(col("a")), max(col("b"))).show()
+---+--------+------+
|  1|first(a)|max(b)|
+---+--------+------+
|  1|       2|     2|
+---+--------+------+
```

it seems `orderBy(lit(1))` directly works, while `groupBy(lit(1))` needs some investigation.

Let me revert this PR first